src/engine/QueryExecutionContext.h

```diff
@@ -22,27 +22,78 @@
 
 class CacheValue {
  private:
-  std::shared_ptr<const Result> _resultTable;
-  RuntimeInformation _runtimeInfo;
+  std::shared_ptr<CacheableResult> resultTable_;
+  RuntimeInformation runtimeInfo_;
+  std::shared_ptr<std::atomic_size_t> currentSize_ =
+      std::make_shared<std::atomic_size_t>(0);
 
  public:
-  explicit CacheValue(Result resultTable, RuntimeInformation runtimeInfo)
-      : _resultTable(std::make_shared<const Result>(std::move(resultTable))),
-        _runtimeInfo(std::move(runtimeInfo)) {}
+  explicit CacheValue(CacheableResult resultTable,
+                      RuntimeInformation runtimeInfo)
+      : resultTable_{std::make_shared<CacheableResult>(std::move(resultTable))},
+        runtimeInfo_{std::move(runtimeInfo)} {
+    if (!resultTable_->isDataEvaluated()) {
+      auto function = resultTable_->resetOnSizeChanged();
+      // We assume this value has previously been set, otherwise this can be
+      // simplified
+      AD_CONTRACT_CHECK(function);
+      resultTable_->setOnSizeChanged(
+          [function = std::move(function), currentSize = currentSize_](
+              bool isShrinkable, bool entryAdded,
+              std::shared_ptr<const IdTable> entry) {
+            ad_utility::MemorySize size = calculateSize(*entry);
+            if (entryAdded) {
+              currentSize->fetch_add(size.getBytes());
+            } else {
+              currentSize->fetch_sub(size.getBytes());
+            }
+            return function(isShrinkable, entryAdded, std::move(entry));
```

Comment:
Thinking about it, I should probably change to code to modify `runtimeInfo_`accordingly